manticore/ethereum.py

```diff
@@ -204,42 +193,172 @@ class DetectIntegerOverflow(Detector):
+        +00000000ffffffff  +0000000000000000  +00000000ffffffff *+3ffffffec0000001 *+7ffffffe80000001 *+800000007fffffff *+bffffffe40000001 *+fffffffe00000001
+
+        '''
+        mul = Operators.SEXTEND(a, 256, 512) * Operators.SEXTEND(b, 256, 512)
+        cond = Operators.UGE(mul, 1 << 256)
+        return cond
 
     def did_evm_execute_instruction_callback(self, state, instruction, arguments, result_ref):
         result = result_ref.value
         mnemonic = instruction.semantics
+        result = result_ref.value
+        ios = False
+        iou = False
 
         if mnemonic == 'ADD':
-            if self._can_add_overflow(state, result, *arguments):
-                self.add_finding_here(state, "Integer overflow at {} instruction".format(mnemonic))
-                if issymbolic(result):
-                    result._taint = result.taint | frozenset(("IOA",))
+            ios = self._signed_add_overflow(state, *arguments)
+            iou = self._unsigned_add_overflow(state, *arguments)
         elif mnemonic == 'MUL':
-            if self._can_mul_overflow(state, result, *arguments):
-                self.add_finding_here(state, "Integer overflow at {} instruction".format(mnemonic))
-                if issymbolic(result):
-                    result._taint = result.taint | frozenset(("IOM",))
+            ios = self._signed_mul_overflow(state, *arguments)
+            iou = self._unsigned_mul_overflow(state, *arguments)
         elif mnemonic == 'SUB':
-            if self._can_sub_underflow(state, *arguments):
-                self.add_finding_here(state, "Integer underflow at {} instruction".format(mnemonic))
-                if issymbolic(result):
-                    result._taint = result.taint | frozenset(("IU",))
-        if mnemonic == 'SSTORE':
-            for arg in arguments:
-                if istainted(arg, 'IOA') or istainted(arg, 'IOM') or istainted(arg, 'IU'):
-                    self.add_finding_here(state, "Result of integuer overflowed intruction is written to the storage")
+            ios = self._signed_sub_overflow(state, *arguments)
+            iou = self._unsigned_sub_overflow(state, *arguments)
+        elif mnemonic == 'SSTORE':
+            where, what = arguments
+            if istainted(what, "SIGNED"):
+                for taint in get_taints(what, "IOS_.*"):
+                    loc = self._get_location(state, taint[4:])
+                    if state.can_be_true(loc[-1]):
+                        self.add_finding(state, *loc[:-1])
+            else:
+                for taint in get_taints(what, "IOU_.*"):
+                    loc = self._get_location(state, taint[4:])
+                    if state.can_be_true(loc[-1]):
+                        self.add_finding(state, *loc[:-1])
+
+        if mnemonic in ('SLT', 'SGT', 'SDIV', 'SMOD'):
+            result = taint_with(result, "SIGNED" % id_val)
```

Comment:
`"SIGNED" % id_val`?

Reply:
Assume it's missing a format string; add it and move it to `.format`? Ditto for below